magefile.go

```diff
@@ -2243,3 +2244,110 @@ func hasCleanOnExit() bool {
 	b, _ := strconv.ParseBool(clean)
 	return b
 }
+
+type dependency struct {
+	Name    string
+	Version string
+}
+
+func (d dependency) Clean(sep string) dependency {
+	cleanFn := func(dep, sep string) string {
+		chunks := strings.SplitN(dep, sep, 2)
+		if len(chunks) == 2 {
+			return chunks[1]
+		}
+
+		return dep
+	}
+
+	return dependency{
+		Name:    cleanFn(d.Name, sep),
+		Version: d.Version,
+	}
+}
+
+func OtelReadme() error {
+	fmt.Println(">> Building internal/pkg/otel/README.md")
+
+	readmeTmpl := filepath.Join("internal", "pkg", "otel", "templates", "README.md.tmpl")
+	readmeOut := filepath.Join("internal", "pkg", "otel", "README.md")
+
+	// read README template
+	tmpl, err := template.ParseFiles(readmeTmpl)
+	if err != nil {
+		return fmt.Errorf("failed to parse README template: %w", err)
+	}
+
+	receivers, exporters, processors, err := getDependencies()
+	if err != nil {
+		return fmt.Errorf("Failed to get OTel dependencies: %w", err)
+	}
+
+	data := struct {
+		Receivers  []dependency
+		Exporters  []dependency
+		Processors []dependency
+	}{
+		Receivers:  receivers,
+		Exporters:  exporters,
+		Processors: processors,
+	}
+
+	// resolve template
+	out, err := os.OpenFile(readmeOut, os.O_CREATE|os.O_TRUNC|os.O_WRONLY, 0644)
+	if err != nil {
+		return fmt.Errorf("failed to open file %s: %w", readmeOut, err)
+	}
+	defer out.Close()
+
+	return tmpl.Execute(out, data)
+}
+
+func getDependencies() (receivers, exporters, processors []dependency, err error) {
+	// read go.mod
+	readFile, err := os.Open("go.mod")
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	scanner := bufio.NewScanner(readFile)
+
+	scanner.Split(bufio.ScanLines)
+	// process imports
+	for scanner.Scan() {
+		l := strings.TrimSpace(scanner.Text())
+		// is otel
+		if !strings.Contains(l, "go.opentelemetry.io/") &&
+			!strings.Contains(l, "github.com/open-telemetry/") {
+			continue
+		}
+
+		parseLine := func(line string) (dependency, error) {
+			chunks := strings.SplitN(line, " ", 2)
+			if len(chunks) != 2 {
+				return dependency{}, fmt.Errorf("incorrect format for line %q", line)
+			}
+			return dependency{
+				Name:    chunks[0],
+				Version: chunks[1],
+			}, nil
+		}
+
+		d, err := parseLine(l)
+		if err != nil {
+			return nil, nil, nil, err
+		}
+
+		if strings.Contains(l, "/receiver/") {
+			receivers = append(receivers, d.Clean("/receiver/"))
+		} else if strings.Contains(l, "/processor/") {
+			processors = append(processors, d.Clean("/processor/"))
+		} else if strings.Contains(l, "/exporter/") {
+			exporters = append(exporters, d.Clean("/exporter/"))
+		}
+	}
+
+	readFile.Close()
```

Comment:
Maybe move this to a `defer` statement right after the the file is opened? That way, we close the file handle even if we return early from this function because of an error.